service/storage_service.cc

```diff
@@ -949,39 +949,57 @@ future<> storage_service::handle_state_normal(inet_address endpoint) {
+
     for (auto t : tokens) {
         // we don't want to update if this node is responsible for the token and it has a later startup time than endpoint.
-        auto current_owner = tmptr->get_endpoint(t);
-        if (!current_owner) {
+        auto current = token_to_endpoint_map.find(t);
+        if (current == token_to_endpoint_map.end()) {
             slogger.debug("handle_state_normal: New node {} at token {}", endpoint, t);
             owned_tokens.insert(t);
-        } else if (endpoint == *current_owner) {
-            slogger.debug("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, *current_owner, t);
+            continue;
+        }
+        auto current_owner = current->second;
+        if (endpoint == current_owner) {
+            slogger.debug("handle_state_normal: endpoint={} == current_owner={} token {}", endpoint, current_owner, t);
             // set state back to normal, since the node may have tried to leave, but failed and is now back up
             owned_tokens.insert(t);
-        } else if (_gossiper.compare_endpoint_startup(endpoint, *current_owner) > 0) {
-            slogger.debug("handle_state_normal: endpoint={} > current_owner={}, token {}", endpoint, *current_owner, t);
+            token_to_endpoint_map.erase(current);
+        } else if (_gossiper.compare_endpoint_startup(endpoint, current_owner) > 0) {
+            slogger.debug("handle_state_normal: endpoint={} > current_owner={}, token {}", endpoint, current_owner, t);
             owned_tokens.insert(t);
-            // currentOwner is no longer current, endpoint is.  Keep track of these moves, because when
-            // a host no longer has any tokens, we'll want to remove it.
-            std::multimap<inet_address, token> ep_to_token_copy = get_token_metadata().get_endpoint_to_token_map_for_reading();
-            auto rg = ep_to_token_copy.equal_range(*current_owner);
-            for (auto it = rg.first; it != rg.second; it++) {
-                if (it->second == t) {
-                    slogger.info("handle_state_normal: remove endpoint={} token={}", *current_owner, t);
-                    ep_to_token_copy.erase(it);
+            slogger.info("handle_state_normal: remove endpoint={} token={}", current_owner, t);
+            token_to_endpoint_map.erase(current);
+            candidates_for_removal.insert(current_owner);
+            slogger.info("handle_state_normal: Nodes {} and {} have the same token {}. {} is the new owner", endpoint, current_owner, t, endpoint);
+        } else {
+            slogger.info("handle_state_normal: Nodes {} and {} have the same token {}. Ignoring {}", endpoint, current_owner, t, endpoint);
+        }
+    }
+
+    // After we replace all tokens owned by current_owner
+    // We check for each candidate for removal if it still owns any tokens,
+    // and remove it if it doesn't anymore.
+    if (!candidates_for_removal.empty()) {
```

Comment:
Shouldn't this commit be merged independently of the rest of the series?

Reply:
It should.  I was a bit lazy since it was easier done on top of of the preliminary patches in an earlier version of the series.  Now I think it should be possible to send it as a standalone patch.

Reply:
Extracted to https://github.com/scylladb/scylladb/pull/12141